raiden/storage/restore.py

```diff
@@ -0,0 +1,80 @@
+from contextlib import contextmanager
+from copy import deepcopy
+
+from raiden.transfer import node, views
+from raiden.transfer.architecture import StateManager
+from raiden.transfer.state import NettingChannelState
+from raiden.transfer.utils import hash_balance_data
+from raiden.utils import typing
+
+
+@contextmanager
+def temporary_state_manager(state_transition, state):
+    state_manager = StateManager(state_transition, state)
+    try:
+        yield state_manager
+    finally:
+        state_manager = None
+
+
+def channel_state_until_balance_hash(
+        raiden: 'RaidenService',
+        token_address: typing.TokenAddress,
+        channel_identifier: typing.ChannelID,
+        target_balance_hash: bytes,
+) -> typing.Optional[NettingChannelState]:
+    """ Go through WAL state changes until a certain balance hash is found. """
+
+    # Restore state from the latest snapshot
+    snapshot = raiden.wal.storage.get_latest_state_snapshot()
```

Comment:
this is not correct, what if the state change which has the right balance proof hash was the before the latest snapshot?